python/ray/data/_internal/planner/plan_write_op.py

```diff
@@ -9,31 +9,50 @@
     MapTransformer,
 )
 from ray.data._internal.logical.operators.write_operator import Write
-from ray.data.block import Block
+from ray.data.block import Block, BlockAccessor
 from ray.data.datasource.datasink import Datasink
 from ray.data.datasource.datasource import Datasource
 
 
 def generate_write_fn(
     datasink_or_legacy_datasource: Union[Datasink, Datasource], **write_args
 ) -> Callable[[Iterator[Block], TaskContext], Iterator[Block]]:
+    def fn(blocks: Iterator[Block], ctx) -> Iterator[Block]:
+        """Writes the blocks to the given datasink or legacy datasource.
+
+        Outputs the original blocks to be written."""
+        if isinstance(datasink_or_legacy_datasource, Datasink):
+            original_blocks = datasink_or_legacy_datasource.write(blocks, ctx)
+        else:
+            original_blocks = datasink_or_legacy_datasource.write(
+                blocks, ctx, **write_args
+            )
+
+        return original_blocks
+
+    return fn
+
+
+def generate_collect_write_stats_fn() -> Callable[
+    [Iterator[Block], TaskContext], Iterator[Block]
+]:
     # If the write op succeeds, the resulting Dataset is a list of
     # arbitrary objects (one object per write task). Otherwise, an error will
     # be raised. The Datasource can handle execution outcomes with the
     # on_write_complete() and on_write_failed().
     def fn(blocks: Iterator[Block], ctx) -> Iterator[Block]:
-        if isinstance(datasink_or_legacy_datasource, Datasink):
-            write_result = datasink_or_legacy_datasource.write(blocks, ctx)
+        """Handles stats collection for block writes."""
+        if all(BlockAccessor.for_block(block).num_rows() == 0 for block in blocks):
+            status = "skip"
         else:
-            write_result = datasink_or_legacy_datasource.write(
-                blocks, ctx, **write_args
-            )
+            status = "ok"
 
         # NOTE: Write tasks can return anything, so we need to wrap it in a valid block
         # type.
         import pandas as pd
 
-        block = pd.DataFrame({"write_result": [write_result]})
+        # TODO(scottjlee): add detailed stats collection for writes.
```

Comment:
@scottjlee can you help me understand how the outputs of this map operator are used? My understanding is that the outputs are only used by `Datasink.on_write_complete`, and aren't related to dataset stats?